### PR TITLE
Refactor code upload worker to spawn single pod for static code upload challenge

### PR DESCRIPTION
**Pull request by savish28**
Context: Static Code Upload Challenge

Deliverables:
- [x] Refactor code upload worker to spawn single pod for static code upload challenge.
- [x] Refactor submission worker to disregard static dataset code upload job submission message.
- [x] Add volume mount setup in job pods to mount test dataset
- [x] Add volume mount setup in job pods to mount scripts for submission.
- [x] Add pre-stop hook to submit the job output csv/json to EvalAI.
- [x] Export stdout, stderr logs from jobs to share with participants
- [x] Create monitor_submission.sh
- [x] Update make_submission.sh to create completed.txt/failed.txt
- [x] Create Sidecar container
- [x] Refactor pre-stop hook code to sidecar container. 

@KhalidRmb 
cc: @Ram81

### CODE REVIEW

File: scripts/workers/code_upload_submission_worker.py
Comment (KhalidRmb) [inline scripts/workers/code_upload_submission_worker.py:149]: @savish28 Please fix the spelling

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:142]: please change this to `life_cycle_config`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:145]: @savish28 is this incomplete?

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:151]: Please rename the container to `submission_container` and name of the container will be `submission`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:153]: environment variable to make submission API call will also be part of this list

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:174]: We can create a new method to get the `Job` object. Parameters of this method will be `submission_pk` with needed specs

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:230]: Let's rename the method to `create_static_code_upload_submission_job_object`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:145]: @Ram81 Initially I thought of creating the pipeline first in this PR and creating the curl request in another PR, but now updating the changes in this PR.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:73]: @savish28 let's move this to a script `make_submission.sh` and in pre-stop hook just do `sh /path/to/make_submission.sh`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:207]: @savish28 `submission` folder should be present when initializing the container

Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:207]: participant can save multiple files we don't need to check if there is only a single file. You can remove the `tail` command. We expect the participant to save submission file as `submission.csv`. If the file is not present we will fail the submission

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:207]: Yeah, I thought, in case the participant deletes the submission folder(by mistake or something) xD our pre-stop hook method might fail, so added "mkdir -p". I guess if we are checking only submission.json or submission.csv I would refactor it and directly check if the file exists.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:117]: @savish28 why are we making curl request from the worker? This is not the approach we discussed. The curl requests will be written in `make_submission.sh` we will only pass environment variables to the container. The curl request will use environment variable. Code upload worker shouldn't substitute curl params and share fixed curl with the submission job.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:147]: please move this curl to `make_submission.sh`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:299]: Here you need to execute `sh /scripts/make_submission.sh` and all required parameters needed for `make_submission.sh` should be environment variables

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:8]: Please add actual curl requests here.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:18]: Add newline

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:237]: @savish28 you can refactor this method to use `get_config_map_volume_object` for dataset volume as well

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:239]: rename it to `evalai-scripts-cm`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:241]: rename to `evalai-scripts`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:247]: let's change the mount path to `evalai_scripts`

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:5]: We need to take this extension as part of challenge config. For now we can use this script. Once we test everything we can switch to accepting submisison extension from challenge config

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:2]: Change this to `Submitting file to EvalAI...`

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:25]: Add a message here `File submission failed` and exit the script after API call.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:28]: Add a message here `File submitted successfully`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:25]: This submission also needs to be deleted from the queue after it is marked failed. Can you please make that change as well.

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:25]: @Ram81, wouldn't the code upload worker do that? I saw there was a condition that deletes the method when the status is changed. 
Current flow: status change -> worker reads the message again -> check status and then deletes.
Is this suggested flow : change status and delete message directly?
Is the idea we also pass receipt_handle and then delete from message by sending it to evalai.

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Done.

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Done.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh]: Done.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:25]: Done.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh]: Done

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Done

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:25]: Yes, the code upload worker will delete it we don't need to do it explicitly

File: scripts/workers/code_upload_submission_worker.py
Comment (KhalidRmb) [inline scripts/workers/code_upload_submission_worker.py:90]: @savish28 Please implement a try catch block here. Always a good practice when you're communicating with something external

File: scripts/workers/code_upload_submission_worker.py
Comment (KhalidRmb) [inline scripts/workers/code_upload_submission_worker.py:98]: @savish28 Please add a try except block + logging message here. Isn't this communicating with something external?

Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:98]: @KhalidRmb this is creating the config map object locally and not connecting to external service at this stage, It connects with eks at a later stage where the actual job is created.

Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:98]: We can take the example of V1Volume as well, all these objects Volume, ConfigMap, VolumeMounts are created as-is, although they are all in a single try-except block at the creation stage where it connects with external client.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:147]: @savish28 this would need to be rebased. See the changes in my recent [PR](https://github.com/Cloud-CV/EvalAI/pull/3497). `busybox` doesn't have `curl` so we have to use a ubuntu image and install curl explicitly. We will merge my PR today, you can rebase after that.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:294]: @savish28 this won't work in a `busybox` container. We have to use the ubuntu image and install curl explicitly. See my PR for reference. Also, the args might also not work. I had to pass the args as part of the command.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:18]: @savish28 instead of this you can use `echo $(date) > "$SUBMISSION_PATH/completed.txt"`. We don't need to create a file explicitly

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/monitor_submission.sh:15]: @savish28 please change the variable name to something like `time_delta` or `wait_interval`

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:13]: Add a log statement here `Submission not complete`

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/monitor_submission.sh:8]: Instead of this we could've just checked for `submission.csv/json` file in the monitor script. We woudn't need a `completed.txt` in that case. Any reason you preferred this setup?

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/monitor_submission.sh:37]: This error is misleading right? This block is executed when submissions times out correct? If that is the case please make `stderr: "Execution time limit exceeded"`. 

We should handle error messages correctly here. If the submission containers exits before creating `submission.csv/json` show `submission.csv not found` otherwise if container is still running after n hours we should `Execution time limit exceeded`.

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:8]: @Ram81 Yes, because I thought monitor_submission.sh should be independent of particulars of what file is submitted. `make_submission.sh` would take care of the particulars. Even if we modify that we want these types of files or maybe in future we want to submit multiple files all that logic would go into `make_submission.sh` and `monitor_submision.sh` would only check for `completed.txt`. In this way we don't have to repeat our logic again and again.

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:15]: @Ram81 do you mean changing `SUBMISSION_TIME_DELTA` to `time_delta`? `next_checkpoint` would be the next time at which we would be going to check.

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:37]: Yeah, we have a plan to check the health of submission docker from sidecar in near future, would add this error there. Changing this error to TLE error as suggested.

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Okay, updated.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:14]: Updated

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:37]: Updated.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh]: Updated.

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:147]: @Ram81, Sure would do the required changes as soon as the respective PR is merged.

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (Ram81) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:15]: I meant change the name of `next_checkpoint` to something relevant to next time interval or similar

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (Ram81) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh:8]: Makes sense.

File: scripts/workers/code_upload_worker_utils/monitor_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/monitor_submission.sh]: Updated.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:560]: Why are we passing `submission_time_limit` in sqs message? This will be a fixed value for each challenge. Let's pass this as task_definition param to the worker. Let's not add too many details in sqs message

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py:560]: @Ram81, Since we are getting challenge object already in code upload worker, why is there a need to pass it from task_defn_worker? 
We could just set an env variable one time at the starting of the code upload worker when we are fetching the challenge details.
OR 
We could have a variable `submission_time_limit` passed explicitly to the function.

The submission_time_limit is not in the actual sqs message though it is in our local message_body variable but yeah we have that separate for good abstraction. The disadvantage is when we have some more info to pass we eventually have to pass it in some dict.
What do you suggest how we proceed?

Comment (Ram81) [reply scripts/workers/code_upload_submission_worker.py:560]: I see, let's create a new dict `submission_meta` and populate that in the same message body object for now. Adding new param to existing functions will need more changes.

Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Updated these changes.

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:96]: This should be full path inside the docker container for it to work when reading the file. This should be `/code/scripts/workers/code_upload_worker_utils/make_submission.sh`

File: scripts/workers/code_upload_submission_worker.py
Comment (Ram81) [inline scripts/workers/code_upload_submission_worker.py:98]: same here

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:3]: @savish28 have you verified these curl requests locally once?

File: tests/unit/challenges/test_views.py
Comment (Ram81) [inline tests/unit/challenges/test_views.py:1864]: Why is this repeated twice here?

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (Ram81) [inline scripts/workers/code_upload_worker_utils/make_submission.sh:2]: @savish28 where are the respective API changes to update the `submission_input_file` and republishing the submission message to the queue? For this API there is not PATCH update available

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: tests/unit/challenges/test_views.py
Comment (savish28) [reply tests/unit/challenges/test_views.py]: Fixed

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:3]: @Ram81 yes.

File: scripts/workers/code_upload_worker_utils/make_submission.sh
Comment (savish28) [reply scripts/workers/code_upload_worker_utils/make_submission.sh:2]: @Ram81 `change_submission_data_and_visibility` in jobs views.py is related to this, we have made changes to send message in this particular view in a previous PR that is merged.

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Done

File: scripts/workers/code_upload_submission_worker.py
Comment (savish28) [reply scripts/workers/code_upload_submission_worker.py]: Updated

**Review by Ram81 — APPROVED** [review]
LGTM

**Review by RishabhJain2018 — APPROVED** [review]
LGTM :)

**Pull request merged**
